server/gfmd/user.c

```diff
@@ -183,13 +160,35 @@ gfarm_error_t
 user_auth_id_modify(struct user *u, const char *auth_id_type,
 	const char *auth_user_id)
 {
-	return (GFARM_ERR_NO_ERROR);
+	enum auth_user_id_type auth_user_id_type = 0;
+	struct db_user_auth_arg arg = {
+		u->ui.username,
+		(char *)auth_id_type,
+		(char *)auth_user_id
+	};
+	gfarm_error_t e;
+
```

Comment:
The user_auth_id_modify() function should update u->auth_user_id[] and user_id_map.
Otherwise, this change notification from the master gfmd (i.e. the case that this function is called from db_journal_apply.c) will not propagate to the slave gfmd (i.e. this gfmd process).

The followings should be done:
* lookup user_id_map to make sure the new auth_user_id is NOT already registered. See GFARM_ERR_ALREDY_EXISTS case in user_enter_gsi_dn().
* remove old auth_user_id value from user_id_map if it's not NULL.
* enter new auth_user_id value to user_id_map and u->auth_user_id[]
* update u->ui.gsi_dn if auth_id_type == "X.509"

When registering to user_id_map, the address of u->user_auth_id[] should be used as auth_user_id key.
The current code has a memory leak problem when trying to delete an entry from the user_id_map.

Note that if this function is called from db_journal_apply.c and an error occurs, gfmd should commit suicide by calling gflog_fatal(, "unknown auth_id_type %s, slave gfmd is older than gfmd", auth_id_type). Because it is a fatal configuration error.
For this reason, we should separate this function into one called from db_journal_apply.c and its subroutine which does actual process, and make the RPC entry point call the subroutine.
The GFARM_ERR_ALREDY_EXISTS case and the GFARM_ERR_NO_MEMORY case should call gflog_fatal() too. (with an appropriate log message)

Also, when u>ui.gsi_dn is changed, u->auth_user_id[GFARM_AUTH_USER_ID_TYPE_X509] should be updated  too.

Also, use_remove_internal() should clear u->auth_user_id[].